java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/sourcenode/common/DbzConnectorConfig.java

```diff
@@ -0,0 +1,133 @@
+    public static final String PG_SCHEMA_NAME = "schema.name";
+
+    private static final String MYSQL_CONFIG_FILE = "mysql.properties";
+    private static final String POSTGRES_CONFIG_FILE = "postgres.properties";
+
+    public static Map<String, String> extractDebeziumProperties(Map<String, String> properties) {
+        // retain only debezium properties if any
+        var userProps = new HashMap<>(properties);
+        userProps.remove(DbzConnectorConfig.HOST);
+        userProps.remove(DbzConnectorConfig.PORT);
+        userProps.remove(DbzConnectorConfig.USER);
+        userProps.remove(DbzConnectorConfig.PASSWORD);
+        userProps.remove(DbzConnectorConfig.DB_NAME);
+        userProps.remove(DbzConnectorConfig.TABLE_NAME);
+        userProps.remove(DbzConnectorConfig.MYSQL_SERVER_ID);
+        userProps.remove(DbzConnectorConfig.PG_SLOT_NAME);
+        userProps.remove(DbzConnectorConfig.PG_SCHEMA_NAME);
+        return userProps;
+    }
+
+    private final long sourceId;
+    private final Properties resolvedDbzProps;
+
+    public long getSourceId() {
+        return sourceId;
+    }
+
+    public Properties getResolvedDebeziumProps() {
+        return resolvedDbzProps;
+    }
+
+    public DbzConnectorConfig(
+            SourceTypeE source, long sourceId, String startOffset, Map<String, String> userProps) {
+        var dbzProps = new Properties();
+        try (var input = getClass().getClassLoader().getResourceAsStream("debezium.properties")) {
+            assert input != null;
+            dbzProps.load(input);
+        } catch (IOException e) {
+            throw new RuntimeException("failed to load debezium.properties", e);
+        }
+
+        StringSubstitutor substitutor = new StringSubstitutor(userProps);
+        if (source == SourceTypeE.MYSQL) {
+            var mysqlProps = initiateDbConfig(MYSQL_CONFIG_FILE, substitutor);
+            // if offset is specified, we will continue binlog reading from the specified offset
+            if (null != startOffset && !startOffset.isBlank()) {
+                // 'snapshot.mode=schema_only_recovery' must be configured if binlog offset is
+                // specified.
+                // It only snapshots the schemas, not the data, and continue binlog reading from the
+                // specified offset
+                mysqlProps.setProperty("snapshot.mode", "schema_only_recovery");
+                mysqlProps.setProperty(
+                        ConfigurableOffsetBackingStore.OFFSET_STATE_VALUE, startOffset);
+            }
+
+            dbzProps.putAll(mysqlProps);
+        } else if (source == SourceTypeE.POSTGRES) {
+            var postgresProps = initiateDbConfig(POSTGRES_CONFIG_FILE, substitutor);
+
+            // if offset is specified, we will continue reading changes from the specified offset
+            if (null != startOffset && !startOffset.isBlank()) {
+                postgresProps.setProperty("snapshot.mode", "never");
```

Comment:
Why not also put this in config file?

Reply:
There is a default config for `snapshot.mode` in config file, but we need to disable snapshotting when `startOffset` is set (e.g. recovery scenario). This logic cannot be achieved by the config file.